### PR TITLE
Units csv master

### PR DESCRIPTION
**Pull request by evertvorster**
This re-unifies the units.csv file, so the tools to edit it works again, enabling us to fix up all the "LOAD FAILED" errors in the HUD much easier.

There was one file that was conflicting with a later edit, modules/missions/privateer.py
It was changed to look exactly like the last version in master now.

Explanation behind this move.
The theory behind cutting up the units.csv file into multiple .csv files is fundamentally flawed. Editing a .csv file, especially one with multiple empty columns by hand is always going to be error-prone. Cutting them up did not make the editing easier. In fact, since the tools to edit that file stopped working, it got much harder and error prone.

### CODE REVIEW

**Comment by nabaco** [comment]
Considering the size of the change, we won't get much from the code review itself.
@evertvorster could you please outline the tests you have done on this change, especially in context of our [PR validation procedure](https://github.com/vegastrike/Vega-Strike-Engine-Source/wiki/Pull-Request-Validation)?

@stephengtuggy @BenjamenMeyer - Any chance you guys could test this locally too, to be on the safe

**Comment by evertvorster** [comment]
Fair enough, this is a big step, and backwards too. 
In this, I understand your concern. 

This commit basically re-sets gameplay to what it was before the units.csv split, as some "balancing" also falls by the wayside. Not sure if it's a bad thing. I like the balance of this commit. 

**Before we get to the PR validation steps:** 
On my local copy of the repo, vegasettings runs fine.

I have a slightly modified vegastrike.settings file from what is in master, to enable 4K. 
Joystick bindings have been modified to enable buttons on my Logiteck Freedom 2.4 Joystick. Swapped the x and z axis to change the feel to more similar to Elite and terrestial aircraft, where if you push the yoke sideways, the craft rolls instead of yaws. My joystick has a dedicated "Twist" which is assigned to yaw. 
Digital HAT switches have been enabled.
Font Point size has been set to 32, to improve game play at 4K. 

**Engine** is 0.6.x branch, compiled to utilize system Python 3 and associated system boost libraries, from Arch AUR. 

Worth noting that saved files from various versions of Assets and engine have "Unspecified" issues when loaded on different Assets and engines. 

**Basic flight.**
Satisfactory. Ship max speed is 120 initially, 125 after purchasing a "Basic  Repair & Refuel"
Controls work as expected. 

**Docking to a planet.**
This works. Auto-docking has never worked for me, and still does not. Even though it's turned on in the settings.

**On planet actions:**
Save the game.
Done, works. Both new saves and overwriting old ones. Generating a new game also works fine. 

**Read news.**
Done on Jacobs, works fine. 

**Buy/Sell some goods.**
Traded to my heart's content. Love this part of the game. Minimal errors in the console, no effect on game play. 

**Buy/Sell a ship upgrade.**
Bought Basic Repair & Upgrade. Error in the console about not finding the unit. No effect on game play. 
Bought Jump Drive. No errors. 

**Accept a mission from the bulletin board.**
Not done in this instance, accepted a mission from Janus. However, I have been accepting missions from the mission computer on other play tests with no issues. 

**Talk to someone in the bar.**
Chatted to Janus. Accepted his mission.

**Primary and secondary weapon usage.**
Both work. Worth noting that on my Joystick mappings of button 2 & 3 have been swapped. Muscle memory of where I expect a missile to be and all. 

**Fight (you can fight with Oswald).**
Done. Oswald is a worthy opponent when your ship is in it's initial trim, it takes some doing to convince him to go away. 
I would say this is the way it is supposed to be. Some things are supposed to be hard.

**SPEC flight.**
SPEC is pretty smooth. Even more smooth than usual. Might be related to the slower ship speed. Auto pilot also works great. 

**A jump to a different system.**
Jumped to Stirling and Bernard's star. No issues, apart from shmrn installations not loading, and not having names. This is fixable with the units.csv file handling tools now working, and kind of the whole point of this PR. The workarounds are trivial, and can be implemented in a day. In the terms of getting a polished 0.6.x out the door, this is achievable in a couple of days. However I would like to get to the bottom of why these things are not loading for a proper fix, and this might take some time. 

**Docking to a non-planet location (mining base, etc).**
Docked with a mining base in Stirling. Works well. 

As an added bonus, I give you:

**Log of gameplay:**

Started a new game, the universe generates without any errors in the console. 
At Atlantis, purchased the basic Repair and Refuel. 
Your craft is limited to 120 whatever speed units it uses before, 125 afterwards. This is one of those balancing changes that fell by the wayside. The jump drive is needed for Janus' mission.
Console complains about "Unit file Basic Repair & Refuel not found", but it works non the less. This is one of the errors that is logged to be fixed, should this PR make the cut.

Bought 118 Malt beverages, they are cheap on Oceanic planets, and I am sure to find a mining base on my way to Jacobs at Bernand's star. Just to stir things up you know, like alcohol is really conductive to a safe mining operation in space. :P

Saved the game. No errors and a nice confirmation screen. 
Went to the bar, accepted Janus' Mission. 
Some chatter in the console about accepting a mission, no ill effects seen on gameplay. There is one warning. 
warning: Unknown tag: python

Saved game again, overwriting the previous savegame. No issues. 

Took off, chatted to Oswald. No issues. Shot at Oswald, pissed him off and had a little skirmish. He flew away. Used SPEC to get to the jump point to Stirling. While in Cephid-17, cycled through all the targets. All there, no missing loads

Console errors at this point:
--------------------------------------------------------------------------------------
 File is a directory !  File is a directory !  File is a directory !  File is a directory ! ship destroyed-no flightgroup
ship destroyed-no flightgroup

ERROR: NULL Unit used in Python script; returning default value...ship destroyed-no flightgroup
cannot execute suborders as Linear Velocity Matcher
-------------------------------------------------------------------------------------------

No effect on gameplay. Jumped to Stirling. 
Console messages at this point:
_____________________________________
ship destroyed-no flightgroup
ship destroyed-no flightgroup
ship destroyed-no flightgroup
cannot execute suborders as Linear Velocity Matcher
ship destroyed-no flightgroup
ship destroyed-no flightgroup
ship destroyed-no flightgroup
ship destroyed-no flightgroup
ship destroyed-no flightgroup
ship destroyed-no flightgroup
ship destroyed-no flightgroup
ship destroyed-no flightgroup
Creating planet Stirling A with texture stars/white_star.texture and technique  - unitlevel <= 2
Creating planet Schmidt with texture planets/Lava.texture and technique  - unitlevel > 2
Unit file factory not found
Creating planet Jump_To_Bernards_star with texture jump.texture and technique  - unitlevel > 2
Creating planet Jump_To_Exile with texture jump.texture and technique  - unitlevel > 2
Creating planet Euphrates with texture planets/Lava.texture and technique  - unitlevel > 2
Unit file research not found
Creating planet Jump_To_Everett with texture jump.texture and technique  - unitlevel > 2
Creating planet Jump_To_Cephid_17 with texture jump.texture and technique  - unitlevel > 2
Creating planet Jump_To_Cardell with texture jump.texture and technique  - unitlevel > 2
Creating planet Angora with texture planets/gas_midget.texture|black.png|planets/gas_midget.texture and technique gas_giants - unitlevel > 2
0(20) : warning C7555: 'varying' is deprecated, use 'in/out' instead
0(21) : warning C7555: 'varying' is deprecated, use 'in/out' instead
0(22) : warning C7555: 'varying' is deprecated, use 'in/out' instead
0(39) : warning C7533: global function texture2DLod is deprecated after version 120
0(61) : warning C7533: global variable gl_LightSource is deprecated after version 120
0(61) : warning C7533: global variable diffuse is deprecated after version 120
0(87) : warning C7533: global function textureCubeLod is deprecated after version 120
0(93) : warning C7533: global variable gl_TexCoord is deprecated after version 120
0(226) : warning C7533: global variable gl_Color is deprecated after version 120
0(243) : warning C7533: global variable gl_FragColor is deprecated after version 120

Fragment info
-------------
0(20) : warning C7555: 'varying' is deprecated, use 'in/out' instead
0(21) : warning C7555: 'varying' is deprecated, use 'in/out' instead
0(22) : warning C7555: 'varying' is deprecated, use 'in/out' instead
0(39) : warning C7533: global function texture2DLod is deprecated after version 120
0(61) : warning C7533: global variable gl_LightSource is deprecated after version 120
0(61) : warning C7533: global variable diffuse is deprecated after version 120
0(87) : warning C7533: global function textureCubeLod is deprecated after version 120
0(93) : warning C7533: global variable gl_TexCoord is deprecated after version 120
0(226) : warning C7533: global variable gl_Color is deprecated after version 120
0(243) : warning C7533: global variable gl_FragColor is deprecated after version 120

Compilation of technique gas_giants successful
Creating planet Latrobe with texture planets/rock1.texture and technique  - unitlevel > 2
Creating planet Jump_To_17-ar with texture jump.texture and technique  - unitlevel > 2
Creating planet Jump_To_Agea with texture jump.texture and technique  - unitlevel > 2
Creating planet Jump_To_Maat with texture jump.texture and technique  - unitlevel > 2
Creating planet Jump_To_Enyo with texture jump.texture and technique  - unitlevel > 2
using NV_CUBE_MAP
Min (0.000000, 0.000000, 0.000000) Max(0.000000, 0.000000, 0.000000) MinLumin 1.000000, MaxLumin 1.000000Read In Star Count 0 used: 2000
Min (0.000000, 0.000000, 0.000000) Max(0.000000, 0.000000, 0.000000) MinLumin 1.000000, MaxLumin 1.000000Read In Star Count 0 used: 38
Loading a starsystem
Loading Star System Crucible/Stirling
 Next To: Crucible/Enyo
 Next To: Crucible/Exile
 Next To: Crucible/Cardell
 Next To: Crucible/Cephid_17
 Next To: Crucible/Everett
 Next To: Crucible/Agea
 Next To: Crucible/Maat
 Next To: Redemption/Bernards_star
 Next To: Crucible/17-ar
Unit is the active player character...changing scene graph
Crucible/Cephid_17  IS NOW AT  Crucible/Stirling
Crucible
Launching bases for Crucible/Stirling
Adding news
ship destroyed-no flightgroup
ship destroyed-no flightgroup
ship destroyed-no flightgroup
ship destroyed-no flightgroup
________________________________________

Notable are these two errors:
Unit file factory not found
Unit file research not found

Accidentally hit Ctrl+C in the console copying this message, game exit. 

Re-started game, loaded last save file, no issues. Popped off-planet, ignored Oswald and SPEC over to the jump point to Stirling. Worth noting that I am seeing almost none of the "jumpy" points on my radars. Switched flight controls to GCNT:TRAVEL to speed things up getting to the jump point.

Jumped to Stirling again. Same errors in console. Cycling through major targets with "N" I see that there are two smaller installations from the shmrn that have a "LOAD FAILED :" error. 

Autopilot over to the mining base in Stirling to go pawn off my booze. Worth noting that the shmrn installations do not have names. I believe that this is an error, as it detracts from the game play. There is no errors about this in the console.

Overshot the mining base horribly as I had a lot of momentum with GCNT:TRAVEL. Switched over to GCNT:MANEUVER and waited for ship to settle down. Auto back to mining base. Approach is now back to it's normal sedate self. 

Docked with mining base. Worth noting that even though "Autodocking" is turned on in settings, it does not. 
Wandered around the mine, looking at all the sights. Sold my booze for a nice 20% profit. Bought Corundum Crystals. 
Console stated that "Unit file Malt_Beverages not found." No effect on gameplay. 

Saved my game, overwriting the previous save. Launched and is auto piloting over to the Jump Point to Bernard's star while writing this. Entering SPEC is nice and smooth. 

While waiting to get close to the jump point, played with the Navigation Computer (Shift+M) Background buttons are still offset. Might be best to just get rid of the buttons totally now, and re-work the interface before the next release.
Moving up to galaxy view and back down to Solar System view all works smoothly, no error messages in the console. 

Jumped to Bernard's star. Nice starfield rendered background. This one I managed to salvage with the high resolution backgrounds. (Not installed on this repo, of course) Mission information "Jump to final system: Bernands's star" turned from olive to bright green. 

Cycled through the major targets, no load errors. The shmrn installations are numbered, not named. No errors on the console. 
Cycled all the minor targets on console, no errors. Autopilot to Jacobs. Ships in this area was a nice mix of factions. dgn, uln, LIHW, purist, shaper, andolian, merchant guild, and so on. Ships are a nice mixture, good to look at. 

Docked with Jacobs. Sold my crystals, again a nice 20% profit. Making money by trading is easy in VS, and in a couple of days IRL you can afford a nice ship. Not much grind, which is a good thing IMHO. Exploring up to Sol is also fun. However, back to the game at hand. 
Console notified me of the completed quest, no indication of that in-game. This is usual, but maybe something we might look at to improve upon in the future. 

Wandered around Jacobs, enjoying the desert sights. Nice. Even managed to grab a bit of news from the dishes out there. Saved game, overwriting the last saved game. Exit Vega Strike from the menu. 

TL;DR : The game works as expected, lovely experience. I hope this counts as a thorough test.

**Comment by nabaco** [comment]
Wow @evertvorster, thanks for the detailed report :)
Indeed it seems that everything works as expected.
My only concern is:
> Worth noting that saved files from various versions of Assets and engine have "Unspecified" issues when loaded on different Assets and engines.

Does that mean that after this change all the existing savefiles won't work and people will have to start the game from the beginning?

**Comment by evertvorster** [comment]
@nabaco, Yeah, I skimmed over the part of the save files. 
Here's what I found so far:
If you have been playing on the Python 2 bindings, and switch over to the Python 3 bindings, there is a pretty good chance that loading the Python 2 save file will core dump the engine. So, unrelated to this issue.

The save files contain the universe, which was created with the split .csv file. 
Luckily only the variable names, and not the layout. So, chances are that a save file from the split .csv files _should_ load, and play with not too many issues. However, there will be installations that do not load. I have some old save files lying around that I can test on. 

The reason I skipped over this detail so quickly is that any change this big is going to be painful. Much like ripping off the band-aid, it's best to get it over quick. Hopefully before the release when loads of people will show up and start playing.

In theory, as I fix up the units.csv file to get rid of the load errors, older savegame files will be better supported.

**Comment by nabaco** [comment]
Hi @evertvorster, I agree that we should do this quick, before the release. But we should at least understand what is the extent of the damage that we are doing.
Best way to do it, IMO, is to get a diff of a new and old savefile, and see if we can write a short script/shell command to do the transition for the end users.
@evertvorster can you try to do at least the initial analysis for this? Or maybe just post here a new and an old version of the savefiles?

@BenjamenMeyer and @Loki1950, I'm bringing this to your attention, as I think we should make this decision together as a team.

**Comment by evertvorster** [comment]
@nabaco I found an ancient save file of mine where I was playing with the Python 3 engine and the split .csv file. It loaded fine, and I was able to fly and trade.

Problem is that every universe that is generated is random, and a diff between one and the other is not going to tell you much. 

As I said, the only issue I see is with the variables contained in the save file. My initial attempts to work around the loading errors are to just have the non-loading unit files in the units.csv file twice, with the requested case for each. There are not THAT many load errors, and as they are reported we can add them to the Units.csv file without too much issue. 

Hackish, I know, but it works. I can use the modding tools to re-generate and fix up any genuinely missing units, until we have a polished product. Once we have the release out the door, then we can tinker with the code to make it less case sensitive and hopefully gain a deeper understanding of it while we are at it. 

At some point in the future I would like to move away from the units.csv file entirely. Preferably with a simple and idiot-proof solution.

**Comment by evertvorster** [comment]
Just FYI, here is the typical Py2 -> Py3 error:
Traceback (most recent call last):
  File "/usr/share/vegastrike/modules/missions/privateer.py", line 34, in Execute
    i.Execute()
  File "/usr/share/vegastrike/modules/random_encounters.py", line 291, in Execute
    dynamic_battle.UpdateCombatTurn()
  File "/usr/share/vegastrike/modules/dynamic_battle.py", line 42, in UpdateCombatTurn
    lookorsiege=LookForTrouble (fac)
  File "/usr/share/vegastrike/modules/dynamic_battle.py", line 352, in LookForTrouble
    i = Director.getSaveString(fg_util.ccp,key,lftiter)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 6: invalid start byte

**Comment by evertvorster** [comment]
@nabaco heheh.... After playing with that save file from the split .csv file, I have a console log, of course. Grepping through that for "Unit file", I get the following list:

Unit file factory not found
Unit file research not found
Unit file mercury not found
Unit file ven not found
Unit file phobos not found
Unit file deimos not found
Unit file jupiter not found
Unit file io not found
Unit file europa not found
Unit file ganymede not found
Unit file callisto not found
Unit file saturn not found
Unit file mimas not found
Unit file enceladus not found
Unit file tethys not found
Unit file dione not found
Unit file rhea not found
Unit file titan not found
Unit file hyperion not found
Unit file iapetus not found
Unit file uranus not found
Unit file miranda not found
Unit file ariel not found
Unit file umbriel not found
Unit file titania not found
Unit file oberon not found
Unit file neptune not found
Unit file proteus not found
Unit file triton not found
Unit file asteroid not found
Unit file pluto not found
Unit file charon not found
Unit file mercury not found
Unit file ven not found
Unit file phobos not found
Unit file deimos not found
Unit file jupiter not found
Unit file io not found
Unit file europa not found
Unit file ganymede not found
Unit file callisto not found
Unit file saturn not found
Unit file mimas not found
Unit file enceladus not found
Unit file tethys not found
Unit file dione not found
Unit file rhea not found
Unit file titan not found
Unit file hyperion not found
Unit file iapetus not found
Unit file uranus not found
Unit file miranda not found
Unit file ariel not found
Unit file umbriel not found
Unit file titania not found
Unit file oberon not found
Unit file neptune not found
Unit file proteus not found
Unit file triton not found
Unit file asteroid not found
Unit file pluto not found
Unit file charon not found
Unit file mercury not found
Unit file ven not found
Unit file phobos not found
Unit file deimos not found
Unit file jupiter not found
Unit file io not found
Unit file europa not found
Unit file ganymede not found
Unit file callisto not found
Unit file saturn not found
Unit file mimas not found
Unit file enceladus not found
Unit file tethys not found
Unit file dione not found
Unit file rhea not found
Unit file titan not found
Unit file hyperion not found
Unit file iapetus not found
Unit file uranus not found
Unit file miranda not found
Unit file ariel not found
Unit file umbriel not found
Unit file titania not found
Unit file oberon not found
Unit file neptune not found
Unit file proteus not found
Unit file triton not found
Unit file asteroid not found
Unit file pluto not found
Unit file charon not found
Unit file engine00 not found
Unit file Koala not found

Aaaand, just fyi:
[evert@Vorster console logs]$ cat 2020-07-08_1.txt |grep "Unit file" | uniq -u | wc -l
98

**Comment by evertvorster** [comment]
Just playing on the Unified units.csv, I only have these few Units not found:
Unit file starfortress not found
Unit file factory not found
Unit file factory not found
Unit file research not found
Unit file engine00 not found
Unit file Corundum_Crystals not found

**Comment by Loki1950** [comment]
Looks like this is going to take some effort.Case sensitivity seems to be biting us in the butt big time I have feeling that this was introduced by some artists being Windows based where case sensitivity just is not an issue then not being tested across platforms.

**Comment by stephengtuggy** [comment]
Hmm. I'm getting nervous about the amount of changes represented here. I think it will take some time to get this all right. And in my opinion, these changes to the assets need to wait until after 0.6.0.

0.6.0 was supposed to be a small release. Just "We Can Release." That's it. And make sure that it's minimally as playable as 0.5.3. IMHO, this PR goes beyond that.

What "should" the initial Llama's speed be? 120? 125? 240? 300? We probably need to decide that together.

What should the joystick mappings be? We probably need to decide that together too. Maybe they should all be options available in `vegasettings`.

The 4K change is fine. I, for one, will be happy to have it. But it should probably go in a separate PR, along with any other changes to `vegastrike.config`.

@evertvorster I would recommend that you use only the Python2 engine for now. The Python3 one is still a work in progress. I believe completion of that is targeted for 0.8.x.

Thank you for all you are doing here, @evertvorster . I just want to make sure we aren't breaking stuff TOO bad in the process of improving the game. :-)

**Comment by evertvorster** [comment]
Have a think all you like. There is no pressure here. If you don't mind, keep the PR open until we decide yay or nay. 

While you are thinking about it, here are some points to consider. 

1. Assets-Production with the split csv files has load errors now. There are things obviously missing from the game when you are flying around, which detracts from game play. If you are happy to release that, fine with me. The Load Errors have been there from before the split, and no one has been able to fix them. (Yet)

2. This PR is not exploring some un-charted territory.  This rolls back the game to the way the game was designed to work up until 2013. Ever wonder why it works so well for me? Since the split, 27 commits were made in total to Assets-Production, most of which were cosmetic. Go have a read through the commits, if you don't believe me.

Off subject, and on to Py 3
I trust the Py 3 engine enough to exclusively play on Py 3, as I don't want to have the hassle of trying to change over my save games later on. If you guys want me to test something in Py2 it's no hassle to swap the engine out quick, but the main engine I will be playing on is Py 3

It will be up to someone who has a Py2 game that he cares about enough the write the converter, or get left behind on Py2.

Just committed a few hours to playing today, by making my pilgrimage from Cephid-17 to Sol, and have already been fixing load errors on my fork, which is what this PR will be until it is merged into master. Of course I'll feed back if I ever find the root cause of these load errors. 

Unfortunately until someone updates the tools to read the split units.csv, my branch will be the only one with load error fixes, unless it's trivial to port.

**Comment by BenjamenMeyer** [comment]
@evertvorster I'm okay reverting the split CSV for the reasons you stated as far as 0.6.x goes, possibly master too. But I'm not sure the team has the required confidence on the engine side for Py3 yet on a number of fronts, and we're going to hold that until 0.8.x becomes master. Right now we just need to understand what we have and be able to release it.

So right now I'm specifically limiting the scope of 0.6.x so we can get it out. All goes well we'll have a quick turnaround and get 0.7.x out quickly too and the be on to 0.8.x.

**Comment by BenjamenMeyer** [comment]
Also I'll review this as soon as I get a chance.

**Comment by evertvorster** [comment]
Fantastic news! Thanks so much for your confidence.

I was playing with 0.6.0 branch on Python 2 for a week or so, and overall it was a very good experience. 

The whole reason that I am not using it now is that Python 3 is already enough for game play, and I get to not worry about upgrading my save file in the future. Any problems we run into, we fix. 

With my Python 2 save file I amassed nearly 2,000,000 credits and explored a healthy chunk of the VS universe. With only a week of my time invested in play so far, I would rather start again and know I have a future-proof save file for the future. There is still so much to do in VS. 

From my point of view, you have been good to go on a release weeks ago. In fact, on Arch Linux at least, you _have_ been released as anyone can install and start to play Vega Strike with this command: Yay -S vegastrike-git

**Comment by nabaco** [comment]
Hi guys,
Let's have the Python3 discussion separately, as it is out of scope of this PR. You can write your comments on that matter [here](https://github.com/vegastrike/Vega-Strike-Engine-Source/issues/76)

Regarding this specific PR: @evertvorster - Do users need to change savefiles due to this specific change in the CSV file? Or the savefile issues are Py2->Py3 specific only?

Other then that, I have no issue merging this PR, as I see it no different from that one:  vegastrike/Vega-Strike-Engine-Source#150. I mean, since the CSV split is quite a "recent" change, and we see that it is causing issues, why not revert it in order to receive a more stable release?
Either way, I think we all agree that we want to get rid of the CSV in the long run, as it proved to be a bad solution for what it tries to solve. So I say, let's merge this, get the release out of the door, and move on to get rid of this dreaded CSV in the future releases.

**Comment by BenjamenMeyer** [comment]
@evertvorster last comment I'm going to make per releases - the big challenge right now for a release is not necessarily the code of 0.6.x but the actual release process, packaging, how users communicate with us, etc. It's more infrastructure related than code related. There's already a lot of folks out there with saved games under Py2; so we'll need to have a converter no matter when we switch to Py3.

Just to be clear: the Units CSV change if accepted to 0.6.x will be the *last* bug fix for 0.6.x that is not release packaging related.

**Comment by evertvorster** [comment]
@Benjamin, if you would like to have some help with the packaging of the project, I can attempt to run the scripts and see if I get an installable package. 
However, as a non-debian or RedHat system I will be using a shim to make it work, and it may not be as valid as a test on a real system. Saying that, there are a few .deb packages in Arch, like Steam, so it's not unheard of. 

The place to start would be to make .deb and .rpm packages available, and if people install them with success, call it a release. Then tell the world. Done. Next item. Easy as that.

**Comment by evertvorster** [comment]
@BenjamenMeyer last comment, then I'll shut up again. 
Make a Python2 package for people that already run the game, and would like to keep their save files. 
It's stable and good, and no effort. (We have it already)

However, also make a Python3 package available for the people just starting out, and willing to accept a small chance of hitting a bug now and then. (It's not like Python2 is bug-free, but enough about that)  Then you run both versions in parallel, and people have a choice. Call the Py3 version 0.6.0-dev.

Let's agree to leave Python 2 in 0.6.x?

**Comment by nabaco** [comment]
@evertvorster back to the specific changes at hand:

> Regarding this specific PR: @evertvorster - Do users need to change savefiles due to this specific change in the CSV file? Or the savefile issues are Py2->Py3 specific only?

Additionally, can you comment on the python changes you did in the PR? They don't seem to be fully related to the CSV change

**Comment by evertvorster** [comment]
Ah, the Python changes, yes. I thought I did make it clear that the edit to that file is to get it to exactly the same as it was in master. From what I can tell, the same file was touched with a later edit with an un-related fix. 

The python edit applies that fix again. TBH, I have not gone into the history of that too much. It was enough for me that someone would go through the effort to put in a fix to see it as a good thing, and worthy of preservation.

**Comment by evertvorster** [comment]
@nabaco - **save files**
There are no apparent issues loading a save file that was generated on a split .csv on to a system with a unified units.csv. 

I have found an example in my own repositories of a savefile generated with the split .csv files and on Python 3. I was able to load this with no issues on the 0.6.0 branch on Python3 that I am currently following. 

This file, which started on a split units.csv repo loaded with no apparent issues on a unified units.csv file. 

If you think about the initial transition from unified to split .csv files, it would make sense to ensure compatibility between the two versions to ease a transition. Because, reasons. 
That very same compatibility should now enable us a smooth transition from split to unified. 

If you are interested, I do have both unified and split repos available on my machine. I can generate a save file and load it on the other to test it in a more controlled test.

**Comment by nabaco** [comment]
@evertvorster loading the savefiles is a good idea.
Otherwise, if we're not expecting issues here with the savefiles, I'm approving this.

**Comment by stephengtuggy** [comment]
My thinking is that past contributors to the assets will get hot under the collar if they find that their changes have been wiped out. (That's happened before.)

I wish that there was a quick and easy way to do this. Unfortunately, I'm not sure that there is, other than going back through those 27 commits since Units.csv was split, and reapplying them to the unified file.

**Comment by evertvorster** [comment]
@stephengtuggy, I have played with these parameters, and the change on the game is negligible. The tool to edit the .csv file safely to apply edits of capability of ships hinges on a unified units.csv.

In this little chicken and egg, the unified unit.csv allows me to make that edit in seconds, and be perfectly safe to the game. If you apply the unified units.csv edit, I will edit the ships to have the same balance as they do now with the included modding tools. I'll stick that in a separate PR, and you can apply as you see fit. I'll also be working on a PR for the loading errors to be applied afterwards. 

Ultimately, I would like to enhance the textures in the game, and these tools make this to be within my reach. High-resolution meshes on installations, anyone?

**Comment by evertvorster** [comment]
@stephengtuggy about past developers getting hot under the collar. 
This is not about anyone's feelings, so let us set that aside. The concern here is about the project. We need the modding tools to work to sort the errors in game play. Since no one is willing to put up the time to fix the modding tools, we have to roll back to a safe place where we have had options. 

The modding tools promise an easy Blender import, and that might be something I can be interested in playing with. If, at the end you get some high resolution textures on installations, would that not be a good thing? Without this unified units.csv edit, those types of goals becomes a little more difficult.

**Comment by stephengtuggy** [comment]
> @stephengtuggy about past developers getting hot under the collar.
> This is not about anyone's feelings, so let us set that aside. The concern here is about the project. We need the modding tools to work to sort the errors in game play. Since no one is willing to put up the time to fix the modding tools, we have to roll back to a safe place where we have had options.
> 

I don't mean that one has to be a slave to everyone's opinions. However, for an open-source project to be healthy, one generally has to have active participation from a number of contributors. To achieve that, people have to **want** to contribute. And to achieve **that,** one generally needs to have at least some regard for people's feelings.

Am I right?

We definitely need someone to get on those modding tools, if only to support the conversion to our [eventual format](https://github.com/vegastrike/Vega-Strike-Engine-Source/issues/169), as soon as we figure out what that is.

For myself, I'm working on the "jumping" bug on master right now.

**Comment by BenjamenMeyer** [comment]
@evertvorster @stephengtuggy  yes we need to be sensitive to folks in that we want people to contribute and encourage that. We also need to make sure that we are moving forward and not backward, so when we revert something we need to make sure it's appropriately targeted to just what needs to get reverted. In this case, there were changes that were done after that need to get re-incorporated and that's a good ask within a PR.

The conversation here is getting very very off-topic. Let's move forward and start keeping conversation on this PR to just what is strictly relevant to the PR; please other side discussions to appropriate Issues or Gitter.

Thanks!

**Comment by evertvorster** [comment]
Agreed, we have gitter for chats. 

The points for and against the PR has been made, unless anyone can think of a reason not to implement this PR that has not been mentioned before? It's pretty clear where I stand on this, as I spent a lot of time and effort to make sure that this PR is stable and safe to apply.

File: modules/trading.py [hunk 1]
```diff
@@ -136,7 +122,7 @@ def cycle(self):
                                         cargo.SetQuantity(quant)
                                         price = prod[1]+vsrandom.uniform(-1,1)*prod[2]
                                         cargo.SetPrice(cargo.GetPrice()*price)
-                                        debug.debug("Adding %s of %s cargo for %s", quant, cargo.GetContent(), price)
+                                        debug.debug("Adding "+str(quant)+" of "+cargo.GetContent()+" cargo for "+str(price))
```
Comment (BenjamenMeyer) [inline modules/trading.py:125]: Looking at the debug module (which we own), then the proper fix should be:
```python
debug.debug(
    "Adding %d of %s for %f", quant. cargo.GetContent(), price)
)
```

Comment (evertvorster) [reply modules/trading.py:125]: @BenjamenMeyer I'm a little lost with the modules/trading.py comment. 

I trust that you know the right thing to do, but I don't know what it is that you want me to do with the information. 

Is this something to be fixed now, in this PR? If it is, how is THAT done? Thanks so much for your patience while I learn the ropes. 

Or, is this something that we'll have to fix with a seperate PR once this one is merged?

**Comment by evertvorster** [comment]
@stephengtuggy About this requested change to this PR. 
Can we do the "Balancing changes" in a seperate PR, please? 
I'm now looking at the "Balancing changes" that were made as a part of the original split in the units.csv commit. 

Honestly, I am having issues the balancing changes, too. Without going into too much detail, it feels like not much thought went into it. Below are charts of vessel speeds. On the left, the vessel speeds from the unified units.cvs. On the right, the speeds from vessels.cvs. 

As you can see, there is quite a bit more variance with the original units.csv. After the split, each class of vessel are now flying the exact same speed. 
How boring is that? What's the point of having a vast array of ships if they are all going to be handling the same?

Imagine all the cars that we buy can only run at exactly the same speed, and handle the same. A show like Top Gear would have been pretty boring. 

If we do this as a separate PR, then we have the option of playing the game as it was intended for a little while, and then applying the "Balancing" changes, or not, as we see fit. There is no issue in transferring the save files. 

I would like to have a chat with the dev that did the balancing changes and get some reasoning behind them. Is he/she still around?

image

File: modules/trading.py (hunk 1 above)
Comment (evertvorster) [reply modules/trading.py:125]: How do I fix this? I do not have a lot of experience with PR yet.

Comment (BenjamenMeyer) [reply modules/trading.py:125]: Since you're reverting we can do it in another PR. Just noting what the correct printf args would be for the parameters.

**Comment by BenjamenMeyer** [comment]
I reviewed last night and most of these changes need to be verified through game play right now, esp given the commit size and our communal knowledge base.

I'll try to run it up this weekend.

**Comment by evertvorster** [comment]
@stephengtuggy, you are totally correct. We can always roll back this PR if it turns out to be a bad decision. 

While I wait for this one to be merged, I'll continue playing on this branch locally. I think I have most of the load errors sorted on my branch now. That PR will be a LOT smaller, I promise!

**Comment by Loki1950** [comment]
Of course it will be smaller just changes in one file ;)

**Comment by nabaco** [comment]
@stephengtuggy can you approve, the PR is blocker currently on you.

**Comment by BenjamenMeyer** [comment]
@nabaco yes the PR doesn't do a `git revert` on purpose. We didn't want to preclude the change from happening in the future.
